### DISCUSSION TITLE
Switch PINs - Question about J2534-2 Protocols : ISO15765_PS , CAN_PS

### DISCUSSION THREAD

**Discussion by tesla20000** [BrianHumlicek/J2534-Sharp#13] [General]
We know that switch J1962_PINS is possible by compatibility with J2534-2 , J2534-1 can't with this. 
In this moment this API have no implemented protocols _PS. 
As is explained in exampled document :
https://beta.quantexlab.com/en/develop/j2534/pt_connect.html
we can see, that possibility to switch between connections 6_14 to 3_11 for CAN is not possible. Missing ports _PS.

If somebody will use this API for example engine, transmission, other modules connected to pins 6+14 in J1962 connector , all will work perfect. 
Question to Brian : 
Is it possible to switch multiplexer pins now? Many J2534 devices now have a multiplexer built in as it is used to communicate with vehicle brands using a different CAN connection in the OBD connector.
How can we solve it?
I hope I'm right and my question is valid.

Details of the _PS protocols are described in SAE J2534_2 in Section 5.3.1
<img width="924" alt="Zrzut ekranu 2020-12-29 o 23 16 08" src="https://user-images.githubusercontent.com/7130668/103317574-d73bc600-4a2b-11eb-856c-433160d8ee59.png">

I found some more info :
[https://beta.quantexlab.com/en/develop/j2534/pt_connect.html](link)

**Comment by BrianHumlicek** [comment]
I will see if I can find the definitions for these and add them to the protocol definitions.  In the meantime, you can probably do...

Device.GetChannel((Protocol)0x8004, Baud.CAN_500000,  ConnectFlag.NONE)

...to connect using CAN_PS.

Try it out.

Brian

**Reply by tesla20000** [reply]
Connection with 0x8004 , or 0x8001 will work, but stil exist problem, I mean API have no function for GET/SET Multiplexer Pin Pair. Thats true ?
Herę You have definitions :

<img width="815" alt="Zrzut ekranu 2020-12-30 o 10 28 15" src="https://user-images.githubusercontent.com/7130668/103342417-bbffa380-4a89-11eb-9e8e-dfd153e48777.png">

<img width="903" alt="Zrzut ekranu 2020-12-30 o 10 29 34" src="https://user-images.githubusercontent.com/7130668/103342511-ea7d7e80-4a89-11eb-8038-6d62f96846b2.png">

// J2534-2 Pin Switched ProtocolIDs
#define J1850VPW_PS			/*-2*/			0x8000 // Not supported
#define J1850PWM_PS			/*-2*/			0x8001 // Not supported
#define ISO9141_PS			/*-2*/			0x8002 // Not supported
#define ISO14230_PS			/*-2*/			0x8003 // Not supported
#define CAN_PS				/*-2*/			0x8004
#define ISO15765_PS			/*-2*/			0x8005
#define J2610_PS			/*-2*/			0x8006 // Not supported
#define SW_ISO15765_PS		/*-2*/			0x8007
#define SW_CAN_PS			/*-2*/			0x8008
#define GM_UART_PS			/*-2*/			0x8009 // Not supported (YET)
#define UART_ECHO_BYTE_PS	/*-2*/			0x800A 
#define HONDA_DIAGH_PS  	/*-2*/			0x800B // Not supported 
#define J1939_PS  	        /*-2*/			0x800C // Not supported
#define J1708_PS  	        /*-2*/			0x800D
#define TP2_0_PS  	        /*-2*/			0x800E // Not supported
#define FT_CAN_PS  	        /*-2*/			0x800F // Not supported
#define FT_ISO15765_PS  	/*-2*/			0x8010 // Not supported

Probably missing is something like this function : 
PassThruIoctl

more about http://www.drewtech.com/support/passthru/ioctl.html

**Comment by BrianHumlicek** [comment]
There is channel.GetConfig and channel.SetConfig which should cover what you need.

Brian

**Reply by tesla20000** [reply]
Hi. Function GetConfig return me error : "Function option is not supported." 
                 In this case is same for CAN , or CAN_PS. No differences when i ask for different Parameter. 
     
                `
                //Protocols - local def :
                const int CAN_PS = 0x8004;
                const int CAN    = 0x0005;
                Channel xx = connManager.initialiseChannel((Protocol)CAN);

                //Parameters :
                const int J1962_PINS = 0x00008001;
                const int ACTIVE_CHANNELS = 0x00008020;
                const int SAMPLE_RATE = 0x00008021;
               
                int paramVal = xx.GetConfig((Parameter)ACTIVE_CHANNELS);
                `
                some suggestions maybe ?
                thanks :)

**Reply by BrianHumlicek** [reply]
The error is from your device, and isn’t anything I can change.

**Reply by BrianHumlicek** [reply]
I started looking at the 2534-2 spec, and it appears that there is a lot more to the implementation than just adding few enum parameters.  I just made the first commit to begin supporting J2534-2.

**Comment by tesla20000** [comment]
It's time to summarize the knowledge about the **STIC SVCI** device.
This device definitely works a bit of a trick. The implementation of PassThru J2534-2 is a bit truncated.

1. Opening the channel for the CAN protocol immediately connects the MUX for pins 6 + 14 - which is OK
2. Opening the channel for the CAN_PS protocol (the protocol was defined only in J2534-2) immediately sets the multiplexer for 3 + 11 - **which is not OK** !!! The J2534 PassThru documentation says the value is 0x0000 (not connected).

For CAN_PS protocol, MUX connection definitions should be defined through Channel.SetConfig (Parameter, Value).
e.g.
 `SetConfig (Parameter.J1962_PINS, PIN_6_14);`
There should be **no default automatic connection** after opening a channel for CAN_PS protocol. This is important.

However, this would not be a problem, if not that STIC SVCI for the CAN_PS protocol allows you to set the value of **ONLY** 0x030B (PIN_3_11), that is the one that is set by default after starting the channel.
Any attempt to set for 0x060E (PIN_6_14) reports the error:
**IOCTL GET_CONFIG / SET_CONFIG parameter value is not recognized.**

Additionally, the SVCI STIC device does not allow to open more than one channel at a time.

Questions for Brian:
1. Do you think it is possible that for the connection 6 + 14 other value than 0x060E can be used?
2. After running the channel.dispose () command, should it be possible to reopen the channel for the same device?

Conclusion.
This device does not implement J2534-1, J2534-2 correctly. It only implements the essentials required by OEM diagnostic software for FORD. Probably other products from China can also pose such problems. Hoping for a 100% working MUX in the device, we should carefully choose Chinese PassThru devices. If we want to use only CAN for pins in the J1962 (OBD2) 6 + 14 connector, STIC SVCI will work smoothly, quickly and will ensure stable communication.

**Comment by tesla20000** [comment]
I can confirm , Your last commit fix problem with J2534-2 support. All fine . Can You update this project in nuget packet manager please ? 
Thanks, Regards

**Reply by BrianHumlicek** [reply]
I was working on a large commit with breaking changes.  I will see if I can finish it off without breaking changes and update the package in the mean time.

**Reply by BrianHumlicek** [reply]
Updated

**Comment by tesla20000** [comment]
Important information. Non-compliance by manufacturers of passthru devices fully to the J2534-2 standard causes that despite the execution of channel.setConfig ((Parameter) 0x8001,0x060E) - correct multiplexer space - some devices during getConfig ((Parameter) 0x8001) still return the value of pin = 0.
As much as we can control the multiplexer, we still cannot check its position / switching efficiency.
As you can see, the creators of Passthr devices do not care about the guidelines of the standards, which makes working with such not fully compatible devices becomes tedious.

Example - AVDI Abrites - Original (no cloned) HW

**Comment by rageart19871** [comment]
I need help setting up a GMLAN 33333 connection. I ask for help in setting up the connection
using SAE.J2534;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Net.NetworkInformation;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace J25SW
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            MessageFilter FlowControlFilter = new MessageFilter()
            {
                FilterType = Filter.FLOW_CONTROL_FILTER,
                Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF },
                Pattern = new byte[] { 0x00, 0x00, 0x07, 0xE8 },
                FlowControl = new byte[] { 0x00, 0x00, 0x07, 0xE0 },
                //Parameter.J1962_PINS,1
                SConfig(Parameter.J1962_PINS, 1)

        };

            string DllFileName = APIFactory.GetAPIinfo().First().Filename;

            using (API API = APIFactory.GetAPI(DllFileName))
            using (Device Device = API.GetDevice())

            using (Channel Channel = Device.GetChannel(Protocol.SW_CAN_PS, Baud.GMLAN_33333, ConnectFlag.CAN_29BIT_ID))
                
            {
                Channel.StartMsgFilter(FlowControlFilter);
                Console.WriteLine($"Voltage is {Channel.MeasureBatteryVoltage() / 1000}");
                
                Channel.SendMessage(new byte[] { 0x00, 0x00, 0x07, 0xE0, 0x01, 0x00 });
                GetMessageResults Response = Channel.GetMessage();
            }
        }
    }
}

**Reply by BrianHumlicek** [reply]
I'm not very familiar with the GM CAN protocols.  I believe there are flags you have to set with the filter and channel setup and possibly with the messages to make it work.  I'll see if I can find more you later when I have a chance.

Brian

**Comment by 33ev** [comment]
Try, 
.GetChanne(Protocol.SW_CAN_PS, Baud.CAN_33333, ConnectFlag.NONE, false); 
.SetConfig(Parameter.J1962_PINS, 0x100);

it works for me.

R

**Reply by rageart19871** [reply]
Which library do you use?

**Reply by 33ev** [reply]
last [1.0.0-CI00036]

**Reply by rageart19871** [reply]
Can you help with the full code for SWCAN?

**Reply by rageart19871** [reply]
private void button1_Click(object sender, EventArgs e)
{
    MessageFilter FlowControlFilter = new MessageFilter()
    {
        FilterType = Filter.FLOW_CONTROL_FILTER,
        Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF },
        Pattern = new byte[] { 0x00, 0x00, 0x06, 0x4C },
        FlowControl = new byte[] { 0x00, 0x00, 0x02, 0x4C },
        

    };

    string DllFileName = APIFactory.GetAPIinfo().First().Filename;

    using (API API = APIFactory.GetAPI(DllFileName))
    using (Device Device = API.GetDevice())

    using (Channel Channel = Device.GetChannel(Protocol.SW_CAN_PS, Baud.GMLAN_33333, ConnectFlag.NONE, false))
    using (SetConfig(Parameter.J1962_PINS, 0x100)) ;

    {
        /*Get.Channel(Protocol.SW_CAN_PS, Baud.GMLAN_33333, ConnectFlag.NONE, false)
        Channel(Protocol.SW_CAN_PS, Baud.GMLAN_33333, ConnectFlag.NONE, false);*/
        
        Channel.StartMsgFilter(FlowControlFilter);
        Console.WriteLine($"Voltage is {Channel.MeasureBatteryVoltage() / 1000}");

        Channel.SendMessage(new byte[] { 0x00, 0x00, 0x02, 0x4C, 0x1A, 0x90 });
        GetMessageResults Response = Channel.GetMessage();
    }
}

**Reply by rageart1987** [reply]
could you help me with the code for swcan gmlan? I'm willing to pay for it.

**Reply by 33ev** [reply]
I don't see "GMLAN_33333" in enum SAEJ2534.Baud, try CAN_33333;
in "SendMessage(new byte[] { 0x00, 0x00, 0x02, 0x4C, 0x1A, 0x90 });" try use all 12 byte 

it tested with china MDI and SM2.

**Comment by tesla20000** [comment]
Not all PassThru devices will support this of course.  This is according J2534-2 full support.  Most china cheap hardware only implement pins 6/14 and 3/11.  Be sure, that hardware can realise this. 
Is important to be sure, that set proper MUX will give you valid response status. I think.

**Reply by rageart19871** [reply]
MDI and mongoose gm full support SWCAN

**Reply by rageart19871** [reply]
private void button1_Click(object sender, EventArgs e)
{
    MessageFilter FlowControlFilter = new MessageFilter()
    {
        FilterType = Filter.FLOW_CONTROL_FILTER,
        Mask = new byte[] { 0xFF, 0xFF, 0xFF, 0xFF },
        Pattern = new byte[] { 0x00, 0x00, 0x06, 0x4C },
        FlowControl = new byte[] { 0x00, 0x00, 0x02, 0x4C },
        

    };

    string DllFileName = APIFactory.GetAPIinfo().First().Filename;

    using (API API = APIFactory.GetAPI(DllFileName))
    using (Device Device = API.GetDevice())

    using (Channel Channel = Device.GetChannel(Protocol.SW_CAN_PS, Baud.GMLAN_33333, ConnectFlag.NONE))
    

    {
        Channel(Protocol.SW_CAN_PS, Baud.GMLAN_33333, ConnectFlag.NONE, false);
        .SetConfig(Parameter.J1962_PINS, 0x100);
        Channel.StartMsgFilter(FlowControlFilter);
        Console.WriteLine($"Voltage is {Channel.MeasureBatteryVoltage() / 1000}");

        Channel.SendMessage(new byte[] { 0x00, 0x00, 0x02, 0x4C, 0x1A, 0x90 });
        GetMessageResults Response = Channel.GetMessage();
    }
}

**Comment by tesla20000** [comment]
- ...fileName.first will give You no guaranty about proper .dll , if You have installed more than one.
next
- For checking voltage I recommend use try..catch. some cloned hardware crash on this place. I know this after test many many PassThru hardwares. 
- Maybe You can try in same time observe CAN line with can bus analyser - to be sure, where is problem ?

**Reply by rageart19871** [reply]
Hi! Please help, communication KWP200(ISO14230) Filters
